poly/src/evaluations/multivariate/multilinear/dense.rs

```diff
@@ -0,0 +1,400 @@
+#[cfg(feature = "parallel")]
+use rayon::prelude::*;
+
+/// Stores a multilinear polynomial in dense evaluation form.
+#[derive(Clone, PartialEq, Eq, Hash, Default, CanonicalSerialize, CanonicalDeserialize)]
+pub struct DenseMultilinearExtension<F: Field> {
+    /// The evaluation over {0,1}^n
+    pub evaluations: Vec<F>,
+    /// Number of variables
+    pub num_vars: usize,
+}
+
+impl<F: Field> DenseMultilinearExtension<F> {
+    /// Construct a new polynomial from a list of evaluations where the index represent a point.
+    ///
+    /// Index represents a point in {0,1}^`num_vars` in little endian form. For example, `0b1011` represents `P(1,1,0,1)`
+    pub fn from_evaluations_slice(num_vars: usize, evaluations: &[F]) -> Self {
+        Self::from_evaluations_vec(num_vars, evaluations.to_vec())
+    }
+
+    /// Construct a new polynomial from a list of evaluations where the index represent a point.
+    ///
+    /// Index represents a point in {0,1}^`num_vars` in little endian form. For example, `0b1011` represents `P(1,1,0,1)`
+    pub fn from_evaluations_vec(num_vars: usize, evaluations: Vec<F>) -> Self {
+        // assert the number of variables match size of evaluations
+        assert_eq!(
+            evaluations.len(),
+            1 << num_vars,
+            "The size of evaluations should be 2 ^ num_vars."
+        );
+
+        Self {
+            num_vars,
+            evaluations,
+        }
+    }
+    /// Relabel the point inplace by switching `k` scalars from position `a` to position `b`, and from position `b` to position `a` in vector.
+    ///
+    /// This function turns `P(x_1,...,x_a,...,x_{a+k - 1},...,x_b,...,x_{b+k - 1},...,x_n)`
+    /// to `P(x_1,...,x_b,...,x_{b+k - 1},...,x_a,...,x_{a+k - 1},...,x_n)`
+    pub fn relabel_inplace(&mut self, mut a: usize, mut b: usize, k: usize) {
+        // enforce order of a and b
+        if a > b {
+            let t = a;
+            a = b;
+            b = t;
+        }
+        assert!(
+            a + k < self.num_vars && b + k < self.num_vars,
+            "invalid relabel argument"
+        );
+        if a == b || k == 0 {
+            return;
+        }
+        assert!(a + k <= b, "overlapped swap window is not allowed");
+        for i in 0..self.evaluations.len() {
+            let j = swap_bits(i, a, b, k);
+            if i < j {
+                self.evaluations.swap(i, j);
+            }
+        }
+    }
+
+    /// returns an iterator that iterates over the evaluations over {0,1}^`num_vars`
+    pub fn iter(&self) -> Iter<'_, F> {
+        self.evaluations.iter()
+    }
+
+    /// returns a mutable iterator that iterates over the evaluations over {0,1}^`num_vars`
+    pub fn iter_mut(&mut self) -> IterMut<'_, F> {
+        self.evaluations.iter_mut()
+    }
+}
+
+impl<F: Field> MultilinearExtension<F> for DenseMultilinearExtension<F> {
+    fn num_vars(&self) -> usize {
+        self.num_vars
+    }
+
+    fn evaluate(&self, point: &[F]) -> F {
```

Comment:
Maybe we should change this to output `Option<F>` if the number of variables doesn't match?

Reply:
yes.